### PR TITLE
tcp_proxy: ignore transfer encoding in HTTP/1.1 CONNECT responses

### LINKED ISSUE #11308: CONNECT support for HTTP/1.1 upstreams
*Description*:

I attempted to configure the second scenario from #1451, where Envoy accepts a raw TCP connection and uses HTTP CONNECT to establish a tunnel through an upstream proxy. This currently works fine when the upstream proxy is also Envoy (e.g. connecting `envoy --config-path configs/encapsulate_in_connect.v3.yaml --base-id 1` to `envoy --config-path configs/terminate_connect.v3.yaml` using the sample YAML files), as the Envoy proxies talk HTTP/2 to each other, but it does not work when the upstream proxy only supports HTTP/1.1 as mentioned on #1451:

> Client --[plain HTTP]--> Envoy --[HTTP CONNECT]--> Upstream Proxy --[HTTPS]--> Server
> We use some upstream non-envoy servers that only support HTTP/1 CONNECT, so HTTP/2 CONNECT wouldn't work for our use case.

It appears that removing `http2_protocol_options` from the cluster configuration is not sufficient to disable HTTP/2 when `tunneling_config` has been specified in the listener's TCP proxy filter. This matches the current .proto docs for `TunnelingConfig`:
> Currently, only HTTP/2 is supported

Raising this as a separate feature request as it seems like the HTTP/2 support is working fine.

### PR DESCRIPTION
**Pull request by irozzo-1A**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: Ignore the transfer encoding header in CONNECT responses
Additional Description: NONE
Risk Level: low
Testing: integration test
Docs Changes: NONE
Release Notes: https://github.com/irozzo-1A/envoy/blob/ignore-transfer-encoding/docs/root/version_history/current.rst#new-features
Platform Specific Features: NONE
Fixes #11308

### CODE REVIEW

**Comment by irozzo-1A** [comment]
xref: https://github.com/envoyproxy/envoy/pull/13293#discussion_r520351145

**Comment by irozzo-1A** [comment]
Hi @alyssawilk, this should be the last piece regarding #11308.

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -102,6 +102,7 @@ New Features
 * signal: added an extension point for custom actions to run on the thread that has encountered a fatal error. Actions are configurable via :ref:`fatal_actions <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.fatal_actions>`.
 * start_tls: :ref:`transport socket<envoy_v3_api_msg_extensions.transport_sockets.starttls.v3.StartTlsConfig>` which starts in clear-text but may programatically be converted to use tls.
 * tcp: added a new :ref:`envoy.overload_actions.reject_incoming_connections <config_overload_manager_overload_actions>` action to reject incoming TCP connections.
+* tcp_proxy: add support for converting raw TCP streams into HTTP/1.1 CONNECT requests. See :ref:`upgrade documentation <tunneling-tcp-over-http>` for details.
```
Comment (alyssawilk) [inline docs/root/version_history/current.rst:105]: wohoo!! :-)

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -1213,13 +1213,12 @@ Envoy::StatusOr<int> ClientConnectionImpl::onHeadersComplete() {
...
-      // For responses to connect requests, do not accept the chunked
-      // encoding header: https://tools.ietf.org/html/rfc7231#section-4.3.6
+      // For responses to connect requests, ignore the chunked encoding header:
+      // https://tools.ietf.org/html/rfc7231#section-4.3.6
       if (headers->TransferEncoding() &&
           absl::EqualsIgnoreCase(headers->TransferEncoding()->value().getStringView(),
                                  Headers::get().TransferEncodingValues.Chunked)) {
-        RETURN_IF_ERROR(sendProtocolError(Http1ResponseCodeDetails::get().InvalidTransferEncoding));
-        return codecProtocolError("http/1.1 protocol error: unsupported transfer encoding");
+        ENVOY_CONN_LOG(trace, "ignoring transfer-encoding in CONNECT response.", connection_);
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc:1221]: I'd be inclined to not even trace log this, but 100% your call.

**Review by alyssawilk — APPROVED** [review]

File: source/common/http/http1/codec_impl.cc (hunk 2 above)
Comment (irozzo-1A) [reply source/common/http/http1/codec_impl.cc]: I'm ok to remove

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**